### ISSUE TITLE
Liquibase lib upgrade to 4.21.1 breaking rollback working earlier

### ISSUE THREAD

**Issue by sachin234s** [liquibase/liquibase#4406]
### Search first

- [X] I searched and no similar issues were found

### Description

Hi Team,

Recently we had updated the liquibase library from 4.8.0 to 4.21.1 and we are noticing the below issue when we are executing liquibase.rollback on a single changeset which is present externally

rror updating database: liquibase.exception.ChangeLogParseException: The file tracing_data_15.xml was not found in the configured search path:
More locations can be added with the 'searchPath' parameter.
liquibase.exception.CommandExecutionException: liquibase.exception.ChangeLogParseException: The file tracing_data_15.xml was not found in the configured search path:
More locations can be added with the 'searchPath' parameter.
	at liquibase.command.CommandScope.execute(CommandScope.java:235)
	at liquibase.Liquibase.rollback(Liquibase.java:734)
	at liquibase.Liquibase.rollback(Liquibase.java:717)
	at liquibase.Liquibase.rollback(Liquibase.java:712)
	at liquibase.Liquibase.rollback(Liquibase.java:696)
	at TestProg3.main(TestProg3.java:44)
Caused by: liquibase.exception.ChangeLogParseException: The file tracing_data_15.xml was not found in the configured search path:
More locations can be added with the 'searchPath' parameter.
	at liquibase.parser.core.xml.XMLChangeLogSAXParser.parseToNode(XMLChangeLogSAXParser.java:114)
	at liquibase.parser.core.xml.AbstractChangeLogParser.parse(AbstractChangeLogParser.java:15)
	at liquibase.command.core.helpers.DatabaseChangelogCommandStep.getDatabaseChangeLog(DatabaseChangelogCommandStep.java:100)
	at liquibase.command.core.helpers.DatabaseChangelogCommandStep.run(DatabaseChangelogCommandStep.java:77)
	at liquibase.command.CommandScope.execute(CommandScope.java:212)
	... 5 more

he code is simple liquibase.rollback below which used to work with earlier version of library

```
String tracingFilePath = "tracxing_data_15.xml";
            SearchPathResourceAccessor resourceAccessor = new SearchPathResourceAccessor(
                    new DirectoryResourceAccessor(Paths.get("C:\\SIU_1011_RC3\\var\\").toAbsolutePath().toFile()));
            Liquibase liquibase = new Liquibase(tracingFilePath, resourceAccessor, database);
            liquibase.validate();
            liquibase.rollback("tracing_data_15_create", "");
```

### Steps To Reproduce

In windowsplace the tracing_data_15.xml in C:\\SIU_1011_RC3\\var\\ dir

place the lib with below code and dependencies in C:\\SIU_1011_RC3\\ directory
String tracingFilePath = "tracxing_data_15.xml";
            SearchPathResourceAccessor resourceAccessor = new SearchPathResourceAccessor(
                    new DirectoryResourceAccessor(Paths.get("C:\\SIU_1011_RC3\\var\\").toAbsolutePath().toFile()));
            Liquibase liquibase = new Liquibase(tracingFilePath, resourceAccessor, database);
            liquibase.validate();
            liquibase.rollback("tracing_data_15_create", "");

Running the above jar fails with liquibase 4.21.1 library with below exception
More locations can be added with the 'searchPath' parameter.
liquibase.exception.CommandExecutionException: liquibase.exception.ChangeLogParseException: The file tracing_data_15.xml was not found in the configured search path:
More locations can be added with the 'searchPath' parameter.
	at liquibase.command.CommandScope.execute(CommandScope.java:235)
	at liquibase.Liquibase.rollback(Liquibase.java:734)
	at liquibase.Liquibase.rollback(Liquibase.java:717)
	at liquibase.Liquibase.rollback(Liquibase.java:712)
	at liquibase.Liquibase.rollback(Liquibase.java:696)
	at TestProg3.main(TestProg3.java:44)
Caused by: liquibase.exception.ChangeLogParseException: The file tracing_data_15.xml was not found in the configured search path:
More locations can be added with the 'searchPath' parameter.
	at liquibase.parser.core.xml.XMLChangeLogSAXParser.parseToNode(XMLChangeLogSAXParser.java:114)
	at liquibase.parser.core.xml.AbstractChangeLogParser.parse(AbstractChangeLogParser.java:15)
	at liquibase.command.core.helpers.DatabaseChangelogCommandStep.getDatabaseChangeLog(DatabaseChangelogCommandStep.java:100)
	at liquibase.command.core.helpers.DatabaseChangelogCommandStep.run(DatabaseChangelogCommandStep.java:77)
	at liquibase.command.CommandScope.execute(CommandScope.java:212)
	... 5 more

The code is simple liquibase.rollback below which used to work with earlier version of library

However the code succeeds with liquibase 4.8.0 library.

### Expected/Desired Behavior

the rollback should succeed ideally

### Liquibase Version

4.21.1

### Database Vendor & Version

mysql 8.0.23

### Liquibase Extensions

liquibase-core

### OS and/or Infrastructure Type/Provider

windows

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR (Thank you!)

**Comment by sachin234s** [comment]
After analysing the liquibase-core source code, i was able to fix the rollback issue as below

before

```
public void rollback(String tagToRollBackTo, String rollbackScript, Contexts contexts,
                         LabelExpression labelExpression) throws LiquibaseException {
        new CommandScope(RollbackCommandStep.COMMAND_NAME)
                .addArgumentValue(DbUrlConnectionCommandStep.DATABASE_ARG, Liquibase.this.getDatabase())
                .addArgumentValue(DatabaseChangelogCommandStep.CHANGELOG_FILE_ARG, changeLogFile)
                .addArgumentValue(DatabaseChangelogCommandStep.CONTEXTS_ARG, (contexts != null? contexts.toString() : null))
                .addArgumentValue(DatabaseChangelogCommandStep.LABEL_FILTER_ARG, (labelExpression != null ? labelExpression.getOriginalString() : null))
                .addArgumentValue(ChangeExecListenerCommandStep.CHANGE_EXEC_LISTENER_ARG, changeExecListener)
                .addArgumentValue(RollbackCommandStep.TAG_ARG, tagToRollBackTo)
                .addArgumentValue(RollbackCommandStep.ROLLBACK_SCRIPT_ARG, rollbackScript)
                .execute();
    }
```

Root cause seems to be as below
1. Rollback always overrode the resourceaccessor to ClassLoaderResourceAccessor, hence any other resource accessor set to liquibase constructor is ignared.
2. this led to the exception below
  **quibase.exception.CommandExecutionException: liquibase.exception.ChangeLogParseException: The file tracing_data_15.xml was not found in the configured search path:**

now the fix is as below
```
public void rollback(String tagToRollBackTo, String rollbackScript, Contexts contexts,
                             LabelExpression labelExpression) throws LiquibaseException {
            runInScope(() -> {
                CommandScope rollbackCommand =new CommandScope(RollbackCommandStep.COMMAND_NAME);
                rollbackCommand.addArgumentValue(DbUrlConnectionCommandStep.DATABASE_ARG, LiquibaseImpl.this.getDatabase());
                rollbackCommand.addArgumentValue(DatabaseChangelogCommandStep.CHANGELOG_FILE_ARG, changeLogFile);
                rollbackCommand.addArgumentValue(DatabaseChangelogCommandStep.CONTEXTS_ARG, (contexts != null ? contexts.toString() : null));
                rollbackCommand.addArgumentValue(DatabaseChangelogCommandStep.LABEL_FILTER_ARG, (labelExpression != null ? labelExpression.getOriginalString() : null));
                rollbackCommand.addArgumentValue(ChangeExecListenerCommandStep.CHANGE_EXEC_LISTENER_ARG, changeExecListener);
                rollbackCommand.addArgumentValue(RollbackCommandStep.TAG_ARG, tagToRollBackTo);
                rollbackCommand.addArgumentValue(RollbackCommandStep.ROLLBACK_SCRIPT_ARG, rollbackScript);
                rollbackCommand.execute();
            });
        }
```

After the above fix the rollback is succeeding. please review and let provide your feedback

**Comment by jacokok** [comment]
This looks good. Having the exact same issue.

For us it is on change sets where we are using properties from liquibase.properties.

**Comment by filipelautert** [comment]
Nice, thanks for finding the issue!  While it is not fixed in core (and as this method is deprecated) you can use the following snippet in your code to get around that:

```
       SearchPathResourceAccessor resourceAccessor = new SearchPathResourceAccessor(
                new DirectoryResourceAccessor(Paths.get("/var/opt/SIU/").toAbsolutePath().toFile()));

        Map<String, Object> scopeObjects = new HashMap<>();
        scopeObjects.put(Scope.Attr.database.name(), myDatabase);
        scopeObjects.put(Scope.Attr.resourceAccessor.name(), resourceAccessor);

        Scope.child(scopeObjects, () -> {
            new CommandScope(RollbackCommandStep.COMMAND_NAME)
                .addArgumentValue(DbUrlConnectionCommandStep.DATABASE_ARG, Liquibase.this.getDatabase())
                .addArgumentValue(DatabaseChangelogCommandStep.CHANGELOG_FILE_ARG, changeLogFile)
                .addArgumentValue(DatabaseChangelogCommandStep.CONTEXTS_ARG, (contexts != null? contexts.toString() : null))
                .addArgumentValue(DatabaseChangelogCommandStep.LABEL_FILTER_ARG, (labelExpression != null ? labelExpression.getOriginalString() : null))
                .addArgumentValue(ChangeExecListenerCommandStep.CHANGE_EXEC_LISTENER_ARG, changeExecListener)
                .addArgumentValue(RollbackCommandStep.TAG_ARG, tagToRollBackTo)
                .addArgumentValue(RollbackCommandStep.ROLLBACK_SCRIPT_ARG, rollbackScript)
                .execute();
        });
```

**Comment by rdhzl** [comment]
@filipelautert I'm a little confused as to the path forward with this. I've got code not too dissimilar to @sachin234s where I construct a Liquibase object, run a flawed migration and attempt a rollback (testing our developers migrations release to release):

Construct:
`Liquibase liquibase = new liquibase.Liquibase(migrationFileName, new DirectoryResourceAccessor(migrationsDirectory), database);`

Update (succeeds):
`liquibase.update(new Contexts(), new LabelExpression());`

Rollback (fails):
`liquibase.rollback(tag, "");`

With error:
```
liquibase.exception.CommandExecutionException: liquibase.exception.ChangeLogParseException: The file db.changelog-test-error-rollback.xml was not found in the configured search path:
More locations can be added with the 'searchPath' parameter.
```

I see that both the update()/rollback() method versions I'm using are now deprecated (and maybe this was a hacky usage on my part), but is the expectation that the rollback() method, while deprecated, should continue to work in this mode? Otherwise is there documentation I can follow to see where I should head with this?

Thanks for any help.

**Comment by rdhzl** [comment]
> 

Hi @filipelautert , thanks for your response. Unfortunately 4.23.0 doesn't fix things for us.

We have two tests that leverage the now deprecated Liquibase class. The first test forces a rollback to the latest tag in databasechangelog. The intent of that test is to ensure developers are creating changesets that can be automatically rolled back release-to-release.

The second test exercises a problematic changelog file with the intent to exercise our automatic rollback code.

Somewhere in the 4.21 era a change happened that broke how we're using this now deprecated API. It seems like Liquibase is perhaps not honoring the ResourceAccessor we're passing into it. So while it knows the master changelog filename we want it to use it has no context around where the file is located.

Right now I'm trying to rework our code to use the CommandScope-based API instead of the Liquibase. It took me a while to understand that's the path forward from the deprecated Liquibase class. I had to stumble upon some forum posts with examples, and then your earlier post above started making sense: https://forum.liquibase.org/t/4-21-1-is-out-deprecates-liquibase-update/8087 and https://stackoverflow.com/a/76051840.

Our issues are a bit further compounded by a poor decision on our part to not store these files in the classpath (i.e. src/main|test/resources), but rather in a directory, and that's something I'm also trying to revisit as well.

Ultimately if there's a better example of documentation around CommandScope or migrating from Liquibase.java to CommandScope and you can link that it would be appreciated. Thanks!

**Comment by blaghed** [comment]
Hi @filipelautert ,

We are also facing this issue, as far as I can understand it.
A quick browse seems to show that this may have been accidentally caused during a refactor between 4.20.0 and 4.21.0, and then further buried with a much larger refactor of the project's layout during 4.22.0.

With the current latest version, there is a new (much tidier) approach to run liquibase, but it does require setting up the execution correctly, as we can see, for example, on the update command:
https://github.com/liquibase/liquibase/blob/v4.23.0/liquibase-standard/src/main/java/liquibase/Liquibase.java#L311
```java
    public void update(Contexts contexts, LabelExpression labelExpression, Writer output, boolean checkLiquibaseTables)
            throws LiquibaseException {
        runInScope(() -> {
            CommandScope updateCommand = new CommandScope(UpdateSqlCommandStep.COMMAND_NAME);
            updateCommand.addArgumentValue(DbUrlConnectionCommandStep.DATABASE_ARG, getDatabase());
            updateCommand.addArgumentValue(UpdateSqlCommandStep.CHANGELOG_FILE_ARG, changeLogFile);
            updateCommand.addArgumentValue(UpdateSqlCommandStep.CONTEXTS_ARG, contexts != null ? contexts.toString() : null);
            updateCommand.addArgumentValue(UpdateSqlCommandStep.LABEL_FILTER_ARG, labelExpression != null ? labelExpression.getOriginalString() : null);
            updateCommand.addArgumentValue(ChangeExecListenerCommandStep.CHANGE_EXEC_LISTENER_ARG, changeExecListener);
            updateCommand.addArgumentValue(DatabaseChangelogCommandStep.CHANGELOG_PARAMETERS, changeLogParameters); // highlighting this part
            updateCommand.setOutput(new WriterOutputStream(output, GlobalConfiguration.OUTPUT_FILE_ENCODING.getCurrentValue()));
            updateCommand.execute();
        });
    }
```

On the rollback command, you may note that the "`DatabaseChangelogCommandStep.CHANGELOG_PARAMETERS`" parameters are not passed in at all:
https://github.com/liquibase/liquibase/blob/v4.23.0/liquibase-standard/src/main/java/liquibase/Liquibase.java#L488-L500
```java
    public void rollback(int changesToRollback, String rollbackScript, Contexts contexts,
                         LabelExpression labelExpression, Writer output) throws LiquibaseException {
        new CommandScope(RollbackCountSqlCommandStep.COMMAND_NAME)
                .addArgumentValue(DbUrlConnectionCommandStep.DATABASE_ARG, Liquibase.this.getDatabase())
                .addArgumentValue(DatabaseChangelogCommandStep.CHANGELOG_FILE_ARG, changeLogFile)
                .addArgumentValue(DatabaseChangelogCommandStep.CONTEXTS_ARG, (contexts != null? contexts.toString() : null))
                .addArgumentValue(DatabaseChangelogCommandStep.LABEL_FILTER_ARG, (labelExpression != null ? labelExpression.getOriginalString() : null))
                .addArgumentValue(ChangeExecListenerCommandStep.CHANGE_EXEC_LISTENER_ARG, changeExecListener)
                .addArgumentValue(RollbackCountCommandStep.COUNT_ARG, changesToRollback)
                .addArgumentValue(AbstractRollbackCommandStep.ROLLBACK_SCRIPT_ARG, rollbackScript)
                .setOutput(new WriterOutputStream(output, GlobalConfiguration.OUTPUT_FILE_ENCODING.getCurrentValue()))
                .execute();
    }
```

There are plenty of other methods in that class making use of "`addArgumentValue`" that also don't set these parameters, so please be sure to check those as well, but I believe that those are likely correct. Only the several "`rollback`" operations are, imho, mistakenly missing that.

This seems like an honest mistake happening due to the refactors, so likely it just needs a small patch for having a quick win.
I don't mind contributing it myself, but I get the feeling I will trip all over your contribution guidelines.
If you simply don't have time for it yourself, however, I will be happy to do so, so just let me know.

**Comment by rdhzl** [comment]
@blaghed : Great catch! Thanks for your investigation.

@filipelautert I don't want to trample on blaghed's find, but after my most recent post I refactored our LB usage to use code very similar to your [post](https://github.com/liquibase/liquibase/issues/4406#issuecomment-1613692578) above, under the (perhaps false) belief that `Liquibase.java` wasn't really the path forward. 

This fixed our issues, but based on blaghed's post I'm wondering if ultimately I should be waiting for a fix here and continuing to work through Liquibase.java. I'm concerned relying on `Scope.child`/`CommandScope` directly might break us in the future. Any thoughts?

**Comment by filipelautert** [comment]
Thanks @rdhzl  and @blaghed ! The code I added above is mostly a copy from what we did on Liquibase class when we refactored rollback command. So we can create a PR to add the missing CHANGELOG_PARAMETERS . Anyone would be able to do create it so I can fast track  to the next release probabily next week? 

About Liquibase class @rdhzl , it will stay deprecated until Liquibase 5 is released. That said, we have no plans for 5 so far, and Liquibase class may even be refactored again for it, but there is a chance it goes away and the way forward is the Command framework - and if Liquibase class remains it will be just wrappers for commands. So Scope.child and CommandScope are the present and the future .

**Comment by blaghed** [comment]
> Anyone would be able to do create it so I can fast track to the next release probabily next week?

Apologies if it oversteps, but here is a PR suggestion: #4591

**Comment by blaghed** [comment]
Hi again,

Sadly this seems to have missed `v4.23.1`, is there something in particular preventing this bug from being fixed?
So far the fix seems like quite a simple thing, but maybe there is some unseen complication that we can further tackle?

Cheers

**Comment by filipelautert** [comment]
@blaghed we are changing our build process and in the last week or so we were not able to run our full test suite against some PR's, and we need all steps to pass before merging to master . Hopefully it will be fixed still this week so we can get it merged for next release.

**Comment by blaghed** [comment]
Hi @filipelautert ,

Thank you for all the hard work on this project.
Is there anything I can do to facilitate including this fix into a release?

Happy to help.
Cheers!

**Comment by filipelautert** [comment]
Hello @blaghed ! You mean the changes on this PR -> https://github.com/liquibase/liquibase/pull/4591 ? They were included in Liquibase 4.24.0 :)

**Comment by blaghed** [comment]
Oh, of course! I was uselessly searching for "4406" on the release notes 🤦 
Thank you very much!
